File: results/0.5/NITF/nitf-claude-3-5-haiku-20241022.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <hammer/hammer.h>
#include <hammer/parsers.h>

typedef struct {
    HBytes *fhdr;
    int64_t clevel;
    HBytes *stype;
    HBytes *ostaid;
    HBytes *fdt;
    HBytes *ftitle;
    HBytes *fscop;
    int64_t fscpys;
    HBytes *encryp;
} NITFFileHeader;

typedef struct {
    HBytes *itype;
    HBytes *irep;
    HBytes *icat;
    HBytes *icords;
    HBytes *igeolo;
} NITFImageSubheader;

typedef struct {
    NITFFileHeader *file_header;
    NITFImageSubheader *image_subheader;
} NITFFile;

HParsedToken* parse_file_header(void* context) {
    HParser *fhdr = h_token((const uint8_t*)"NITF02.10", 9);
    HParser *clevel = h_int_range(h_ch_range('0', '9'), 0, 9);
    HParser *stype = h_choice(h_token((const uint8_t*)"C", 1), h_token((const uint8_t*)"R", 1), NULL);
    HParser *ostaid = h_repeat_n(h_ch_range('A', 'Z'), 10);
    HParser *fdt = h_repeat_n(h_ch_range('0', '9'), 14);
    HParser *ftitle = h_repeat_n(h_ch_range('A', 'Z'), 80);
    HParser *fscop = h_choice(
        h_token((const uint8_t*)"TOP SECRET", 10),
        h_token((const uint8_t*)"SECRET", 6),
        h_token((const uint8_t*)"CONFIDENTIAL", 12),
        NULL
    );
    HParser *fscpys = h_int_range(h_ch_range('0', '9'), 0, 999999);
    HParser *encryp = h_choice(h_token((const uint8_t*)"0", 1), h_token((const uint8_t*)"1", 1), NULL);

    HParser *file_header = h_sequence(
        fhdr, clevel, stype, ostaid, fdt, 
        ftitle, fscop, fscpys, encryp, NULL
    );

    HParseResult *result = h_parse(file_header, context, 0);
    return result ? (HParsedToken*)result->ast : NULL;
}

HParsedToken* parse_image_subheader(void* context) {
    HParser *itype = h_choice(
        h_token((const uint8_t*)"MONO", 4), 
        h_token((const uint8_t*)"RGB", 3), 
        h_token((const uint8_t*)"MULTI", 5), 
        NULL
    );
    HParser *irep = h_choice(
        h_token((const uint8_t*)"MONO", 4), 
        h_token((const uint8_t*)"RGB", 3), 
        h_token((const uint8_t*)"MULTI", 5), 
        NULL
    );
    HParser *icat = h_choice(
        h_token((const uint8_t*)"VIS", 3), 
        h_token((const uint8_t*)"IR", 2), 
        h_token((const uint8_t*)"SAR", 3), 
        NULL
    );
    HParser *icords = h_choice(
        h_token((const uint8_t*)"N", 1), 
        h_token((const uint8_t*)"S", 1), 
        h_token((const uint8_t*)"D", 1), 
        NULL
    );
    HParser *igeolo = h_repeat_n(h_ch_range('0', '9'), 60);

    HParser *image_subheader = h_sequence(
        itype, irep, icat, icords, igeolo, NULL
    );

    HParseResult *result = h_parse(image_subheader, context, 0);
    return result ? (HParsedToken*)result->ast : NULL;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <nitf_file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t *buffer = malloc(file_size);
    if (!buffer) {
        perror("Memory allocation error");
        fclose(file);
        return 1;
    }

    size_t read_size = fread(buffer, 1, file_size, file);
    if (read_size != file_size) {
        perror("File read error");
        free(buffer);
        fclose(file);
        return 1;
    }

    HParsedToken *file_header_result = parse_file_header(buffer);
    HParsedToken *image_subheader_result = parse_image_subheader(buffer);

    if (file_header_result && image_subheader_result) {
        printf("NITF file parsed successfully\n");
    } else {
        printf("NITF file parsing failed\n");
    }

    free(buffer);
    fclose(file);
    return 0;
}

File: results/0.5/NITF/nitf-claude-3-5-haiku-20241022.rs
use nom::{
    IResult,
    bytes::complete::take,
    number::complete::le_u32,
    multi::many0,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
struct NitfFile {
    header: NitfHeader,
    image_segments: Vec<ImageSegment>,
}

#[derive(Debug)]
struct NitfHeader {
    fhdr: String,
    clevel: u8,
    stype: String,
    ostaid: String,
    fdt: String,
}

#[derive(Debug)]
struct ImageSegment {
    image_header: ImageHeader,
    image_data: Vec<u8>,
}

#[derive(Debug)]
struct ImageHeader {
    icat: String,
    iclod: String,
    encryp: u8,
    icords: String,
}

fn parse_nitf_header(input: &[u8]) -> IResult<&[u8], NitfHeader> {
    let (input, fhdr) = take(9usize)(input)?;
    let (input, clevel) = take(1usize)(input)?;
    let (input, stype) = take(1usize)(input)?;
    let (input, ostaid) = take(10usize)(input)?;
    let (input, fdt) = take(14usize)(input)?;

    Ok((input, NitfHeader {
        fhdr: String::from_utf8_lossy(fhdr).to_string(),
        clevel: clevel[0],
        stype: String::from_utf8_lossy(stype).to_string(),
        ostaid: String::from_utf8_lossy(ostaid).to_string(),
        fdt: String::from_utf8_lossy(fdt).to_string(),
    }))
}

fn parse_image_header(input: &[u8]) -> IResult<&[u8], ImageHeader> {
    let (input, icat) = take(2usize)(input)?;
    let (input, iclod) = take(1usize)(input)?;
    let (input, encryp) = take(1usize)(input)?;
    let (input, icords) = take(1usize)(input)?;

    Ok((input, ImageHeader {
        icat: String::from_utf8_lossy(icat).to_string(),
        iclod: String::from_utf8_lossy(iclod).to_string(),
        encryp: encryp[0],
        icords: String::from_utf8_lossy(icords).to_string(),
    }))
}

fn parse_nitf(input: &[u8]) -> IResult<&[u8], NitfFile> {
    let (input, header) = parse_nitf_header(input)?;
    let (input, image_segments) = many0(parse_image_segment)(input)?;

    Ok((input, NitfFile {
        header,
        image_segments,
    }))
}

fn parse_image_segment(input: &[u8]) -> IResult<&[u8], ImageSegment> {
    let (input, image_header) = parse_image_header(input)?;
    let (input, image_data_size) = le_u32(input)?;
    let (input, image_data) = take(image_data_size as usize)(input)?;

    Ok((input, ImageSegment {
        image_header,
        image_data: image_data.to_vec(),
    }))
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <nitf_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_nitf(&buffer) {
        Ok((_, nitf)) => {
            println!("Parsed NITF: {:?}", nitf);
            Ok(())
        }
        Err(e) => {
            eprintln!("Parsing error: {:?}", e);
            Err(Box::new(std::io::Error::new(std::io::ErrorKind::InvalidData, format!("{:?}", e))))
        }
    }
}